lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <maths/CTreeShapFeatureImportance.h>
+
+#include <algorithm>
+#include <tuple>
+
+namespace ml {
+namespace maths {
+
+using TRowItr = core::CDataFrame::TRowItr;
+
+CTreeShapFeatureImportance::TDoubleVec
+CTreeShapFeatureImportance::shap(core::CDataFrame& frame,
+                                 const CDataFrameCategoryEncoder& encoder,
+                                 std::size_t numberFeatures,
+                                 std::size_t offset) {
+    numberFeatures = (numberFeatures != -1) ? numberFeatures : frame.numberColumns();
+    std::vector<std::size_t> maxDepthVec;
+    maxDepthVec.reserve(m_Trees.size());
+    for (auto& tree : m_Trees) {
+        auto samplesPerNode =
+            CTreeShapFeatureImportance::samplesPerNode(tree, frame, encoder, 0);
```

Comment:
This is passing in 0 for number of threads! But I think the method can just use the member variable in any case and we can drop this parameter.

Reply:
Good catch!